Adds retryAfterSeconds, retryIterator, timeoutSeconds

morph166955 opened the pull request:
Adds an iterator for number of retries before pausing for retryAfterSeconds - Adds #130 

From openhab/openhab-addons#5892

Mentioned in https://community.openhab.org/t/panasonictv-oh3/108124
https://community.openhab.org/t/too-much-time-before-a-sonos-thing-becomes-definitively-online/62214

New setting retryAfterSeconds to manually set pause between attempts.  retryAfterSeconds address issues where the system loses track of a device (because of a variety of reasons) and then only attempts to reestablish connection by default at the 10 minute mark.  This is entirely too long for a device to be offline.  These configuration items allow for a user to specify a reduced time.  While we could dial this down statically, that doesn't provide end users with flexibility.  This allows for them to specify what they need for their specific application if they run into this issue.

New setting retryIterator to manually set number of iterations to execute consecutively before waiting.  In some cases (e.g. Panasonic TVs as seen in https://github.com/openhab/openhab-core/issues/2223) devices may start sending upnp broadcasts on power up before they are ready for actual connection attempts.  This causes jupnp to wait retryAfterSeconds before making a second connection attempt.  By default, that's 10 minutes.  Adding retryIterations makes jupnp retry a few times before pausing for retryAfterSeconds.  Also, adding the configuration of timeoutSeconds allows the end user to specify how long to wait for each iteration to fail.  This gives the end user the ability to adjust the settings based on their environment.

New setting timeoutSeconds to manually set the HTTP timeout

retryAfterSeconds code written by: lolodomo - merged into this branch by morph166955

Signed-off-by: morph166955 <rosenblumb@gmail.com>

morph166955 commented:
Build will fail until #129 is committed.  That should resolve automatically once that's done.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -41,23 +41,39 @@
         // We want to track how long it takes
         long start = System.nanoTime();
+	
+	if (failedTries.get(requestMessage.getUri()) == null) {
```
using `putIfAbsent` would me more elegant here.

morph166955 replied:
Will change in next commit.

morph166955 replied:
Fixed.  This also fixes several issues below.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -41,23 +41,39 @@
-                failedRequests.remove(requestMessage.getUri());
+		log.debug("Clearing failed attempt after {} tries",
+				numberOfTries);
+      	        failedRequests.remove(requestMessage.getUri());
```
something wrong with the indentation here.

morph166955 replied:
Will change in next commit.

morph166955 replied:
Fixed.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -41,23 +41,39 @@
+	    } else if (start - previeousFailureTime < TimeUnit.SECONDS
+                    .toNanos(getConfiguration().getRetryAfterSeconds()) &&
+                    numberOfTries > 0 ) {
+		log.debug("Previous attempt failed {} times.  Will retry {}",
```
something wrong with the indentation here.

morph166955 replied:
Will change in next commit.

morph166955 replied:
Fixed.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -155,6 +171,11 @@ private void handleRequestFailure(StreamRequestMessage requestMessage) {
         final long currentTime = System.nanoTime();
         failedRequests.put(requestMessage.getUri(), currentTime);
 
+	if (failedTries.get(requestMessage.getUri()) == null) {
```
use `putIfAbsent`

morph166955 replied:
Will change in next commit.

morph166955 replied:
Fixed.  This was unnecessary as we use putIfAbsent on line 52.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClientConfiguration.java:
```diff
@@ -63,6 +64,20 @@ public void setTimeoutSeconds(int timeoutSeconds) {
+
+    public void setRetryIterations(int retryIterations) {
+	if (retryIterations < 0) {
+		throw new  IllegalArgumentException("Retry iterations can not be null!");
```
```suggestion
		throw new IllegalArgumentException("Retry iterations can not be null!");
```

kaikreuzer replied:
It isn't null, it is less than 0.

morph166955 replied:
Fair, I'll change the wording.  Will change in next commit.

morph166955 replied:
Fixed

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -155,6 +171,11 @@ private void handleRequestFailure(StreamRequestMessage requestMessage) {
+	    failedTries.put(requestMessage.getUri(), failedTries.get(requestMessage.getUri()) + 1);
+	}
         if (failedRequests.size() > FAILED_REQUESTS_MAX_SIZE) {
             cleanOldFailedRequests(currentTime);
```
Wouldn't something similar also make sense for `failedTries` to prevent the list becoming a resource leak?

morph166955 replied:
I've removed this block of code entirely as it is now redundant.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/TransportConfiguration.java:
```diff
@@ -37,7 +37,7 @@
      * @return created {@link StreamClient}
      */
-    StreamClient<SCC> createStreamClient(final ExecutorService executorService);
+    StreamClient<SCC> createStreamClient(final ExecutorService executorService, int retryAfterSeconds, int retryIterations, int timeoutSeconds);
```
Instead of adding single parameters here, wouldn't it be cleaner to pass a `StreamClientConfiguration`? With that, the method signature could stay untouched, even if additional parameters are added or changed in future.

morph166955 replied:
I hadn't thought about that.  Completely agree.  That will definitely make it easier to add additional options in the future and not require all of this to be touched.  Will change in next commit.

morph166955 replied:
I'll commit this once I can compile this PR cleanly again.  There are a bunch of changes to be made to make this work correctly, I need to be able to compile locally to test it out before committing this.

morph166955 replied:
I've pulled this out and into https://github.com/jupnp/jupnp/pull/134

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClientConfiguration.java:
```diff
@@ -28,6 +28,7 @@
     protected int timeoutSeconds = 10;
     protected int logWarningSeconds = 5;
     protected int retryAfterSeconds = (int) TimeUnit.MINUTES.toSeconds(10);
+    protected int retryIterations = 5;
```
An additional constructor should be added for it as well, since we have it for all the other parameters.

morph166955 replied:
Sorry, not following what I'm missing here.

kaikreuzer replied:
Have a look at lines 33-46 and you should see what I mean.

morph166955 replied:
I went back and looked at the code on this.  We never had a constructor for retryAfterSeconds.  We only have one for timeoutSeconds and logWarningSeconds.  I can easily add the following if you still want this:

    protected AbstractStreamClientConfiguration(ExecutorService requestExecutorService, int timeoutSeconds, int logWarningSeconds, int retryAfterSeconds) {
 
and

    protected AbstractStreamClientConfiguration(ExecutorService requestExecutorService, int timeoutSeconds, int logWarningSeconds, int retryAfterSeconds, int retryIterations) {

Thoughts?

kaikreuzer replied:
I think it would be nice as you could then use this constructor [here](https://github.com/jupnp/jupnp/pull/133/files#diff-337b3d48beb4442659d75e514b5936ab3280418aa9e895d3d13bdc2a56ce9ddfR24) and get rid of the 3 setter calls.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/OSGiUpnpServiceConfiguration.java:
```diff
@@ -86,13 +90,21 @@
+    private boolean mainThreadPool = true;
+    private boolean remoteThreadPool = true;
     private Namespace callbackURI = new Namespace("http://localhost/upnpcallback");
+    private int retryAfterSeconds = -1;
```
It would look better, if we'd also add the default values in all three cases here. Otherwise it gives the impression that this feature is deactivated, which isn't the case.

morph166955 replied:
I believe I did this because of the check implemented here: bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/jetty/JettyTransportConfiguration.java 

The defaults are established in: 
bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClientConfiguration.java 

The logic was effectively "the end user has not chosen to modify the defaults, deliberately send -1 so that we use the defaults."  I can set them to the defaults if you would prefer that instead.

morph166955 replied:
Fixed.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -155,6 +171,11 @@ private void handleRequestFailure(StreamRequestMessage requestMessage) {
+	if (failedTries.get(requestMessage.getUri()) == null) {
+	    failedTries.put(requestMessage.getUri(), (long) 1);
+	} else {
+	    failedTries.put(requestMessage.getUri(), failedTries.get(requestMessage.getUri()) + 1);
```
It seems we are always increasing the value here and only clear it in case of a successful connection. But this means that once we reached the retryIterations, we will be stuck forever. Or am I overlooking something?

morph166955 replied:
We ultimately clear both failedTries and failedRequests after retryAfterSeconds expires above.  Effectively it works like this:

something makes request
failedTries = 0
while failedTries < retryIterations keep repeating request attempt
once failedTries = retryIterations wait retryAfterSeconds (which starts counting at the last attempt)
once retryAfterSeconds expires, reset failedTries and failedRequests and start over again.

This specific code just keeps track of the number of attempts/iterations that have failed.

kaikreuzer replied:
Ok, thanks, I didn't spot the according else branch before.

What is actually making sure that the requests are repeated and how quickly is that done?

If I get it correctly, the current behavior is that it will only do a single try and return null for all others until the timeout is reached, so effectively `retryIterations` is 1 and we now allow it to be increased, is that correct?

morph166955 replied:
sendRequest() is called by the send() funciton in RouterImpl.  I haven't dug into the code to completely understand it but I believe that the requests are a function of several things (scheduled timers, multicast messages, etc) as part of the UPnP spec.

You are correct.  Current behavior is to do a single try (as if retryIterations is set to 1) and then send null until retryAfterSeconds elapses (which right now is 10 minutes by default which is really problematic for some devices).  So we've done two things in this PR.  1) We've made retryAfterSeconds (which has existed for some time now) configurable by the user.  2) We've added retryIterations entirely so that the system makes a few attempts before waiting.  This helps a device come back or online "quickly".  I have seen instances in my own network (with things like my Sonos speakers) where the device just gets bogged down and doesn't reply.  This doesn't mean that the device is dead, just not as responsive as it should be.  Another example is the Panasonic TVs.  From what I can tell (and I don't have one so this is me reading bugs and forum comments), the Panasonic TV sends the UPnP NOTIFY message before the system is actually able to accept SUBSCRIBE messages (which yes is definitely a problem with the TV).  In the current state, that TV won't actually come online for 10 minutes.  Adding retryIterations (and then having the user fine some sane value to fit their needs) hopefully will allow for the TV to come up and work before it fails and goes to retryAfterSeconds.

morph166955 replied:
Am I good to resolve this one?

morph166955 replied:
I reworked this, closing.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -171,6 +192,12 @@ private void handleRequestTimeout(StreamRequestMessage requestMessage, RequestWr
             failedRequests.put(requestMessage.getUri(), currentTime);
         }
 
+        if (failedTries.get(requestMessage.getUri()) == null) {
```
use `putIfAbsent`

morph166955 replied:
Will change in next commit.

morph166955 replied:
Fixed.  This was unnecessary as we use putIfAbsent on line 52.

morph166955 commented:
To note, before committing this, I need to rebase this PR once #129 is finalized.  I really want to do a round of testing on this as I can't really compile it cleanly without #129 being part of the mix.

morph166955 commented:
@kaikreuzer Don't commit this yet, still adjusting it to work with #134 and removing remaining #129 fragments.  I'll post when this is clean.

morph166955 commented:
@kaikreuzer Ok, this builds clean now.  I have NOT tested the jar yet to make sure it all functions but it "looks ok" at this point to me.  Feel free to do a once-over and see if I missed anything glaring.

morph166955 commented:
It occurs to me that since we have StreamConfigurationClient we no longer need to have the actual variables or defaults in the OSGi code.  They will populate by default and only change if we set configs manually.  This simplifies the code.

morph166955 commented:
I ran some tests on this and I was getting NPEs.  Turns out, I missed something with the StreamClientConfiguration.  I submitted https://github.com/jupnp/jupnp/pull/135 to fix the NPE problem.

morph166955 commented:
This tests and works cleanly (with the patch submitted in #135 applied).  

This is an example of it functioning.  I powered off one of my sonos speakers and it caused the failure:

```
2021-03-28 17:16:34.954 [WARN ] [org.jupnp.transport.spi.StreamClient] - HTTP request failed: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
java.net.NoRouteToHostException: No route to host
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779) ~[?:?]
        at org.eclipse.jetty.io.SelectorManager.doFinishConnect(SelectorManager.java:355) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector.processConnect(ManagedSelector.java:223) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector.access$1500(ManagedSelector.java:61) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processSelected(ManagedSelector.java:533) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:395) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) ~[?:?]
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918) ~[?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
2021-03-28 17:16:35.956 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 1 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:37.957 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:38.958 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 2 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:40.958 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:41.960 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 3 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:43.960 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:44.961 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 4 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:46.962 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:47.963 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 5 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:49.963 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:50.964 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 6 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:52.965 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:53.966 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 7 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:55.966 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:56.968 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 8 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:58.968 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:16:59.969 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 9 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:01.969 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:02.970 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:03.971 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:04.972 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:05.973 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:06.973 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:07.974 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:08.975 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:09.975 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:10.976 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:11.977 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Clearing failed attempt after 10 tries
2021-03-28 17:17:13.978 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:14.979 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 1 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:15.049 [WARN ] [org.jupnp.transport.spi.StreamClient] - HTTP request failed: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
java.net.NoRouteToHostException: No route to host
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779) ~[?:?]
        at org.eclipse.jetty.io.SelectorManager.doFinishConnect(SelectorManager.java:355) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector.processConnect(ManagedSelector.java:223) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector.access$1500(ManagedSelector.java:61) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processSelected(ManagedSelector.java:533) ~[?:?]
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:395) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) ~[?:?]
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918) ~[?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
2021-03-28 17:17:16.051 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 2 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:18.052 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:19.053 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 3 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:21.053 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:22.054 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 4 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:24.055 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:25.056 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 5 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:27.056 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:28.057 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 6 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:30.058 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:31.059 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 7 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:33.060 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:34.061 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 8 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:36.061 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:37.062 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 9 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:39.062 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:40.063 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:41.065 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:42.066 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:43.066 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:44.067 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:45.068 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:46.068 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:47.069 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:48.070 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Will not attempt request because it failed 10 times in the last 10 seconds: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:49.070 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Clearing failed attempt after 10 tries
2021-03-28 17:17:51.071 [INFO ] [org.jupnp.transport.spi.StreamClient] - Timeout of 2 seconds while waiting for HTTP request to complete, aborting: (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control
2021-03-28 17:17:52.072 [DEBUG] [org.jupnp.transport.spi.StreamClient] - Previous attempt failed 1 times.  Will retry (OutgoingActionRequestMessage) POST http://REDACTED:1400/DeviceProperties/Control


```

morph166955 commented:
@kaikreuzer Unless you have additional comments, I don't have anything else to commit to my branch.  I believe this now works as expected and is ready to commit.  Thank you!

morph166955 commented:
👀

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/AbstractStreamClient.java:
```diff
@@ -57,16 +58,29 @@ public StreamResponseMessage sendRequest(StreamRequestMessage requestMessage) th
         // We want to track how long it takes
         long start = System.nanoTime();
+	
+	failedTries.putIfAbsent(requestMessage.getUri(), (long) 0);
```
Please check the indentation in that file - there are tabs and spaces mixed.

morph166955 replied:
I think I got them all.  Please let me know if I missed one.

kaikreuzer commented on bundles/org.jupnp/src/main/java/org/jupnp/transport/spi/StreamClientConfiguration.java:
```diff
@@ -47,6 +52,21 @@
+    /**
+     * Sets the HTTP Timeout in Seconds
+     */
+    public void setTimeoutSeconds(int timeoutSeconds);
```
we do not have any setter methods for the other fields, so if you add the additional constructor as suggested above, we could remove these three methods here, which would seem to be cleaner.

morph166955 replied:
We need the setters so that the OSGi configs can be set as part of creating the StreamClientConfiguration that we pass to createStreamClient in OSGiUpnpServiceConfiguration.java.  So we can definitely still use the constructor in JettyTransportConfiguration.java but we still need the setter for OSGi.

kaikreuzer replied:
Hm, this does not look clean to me. A configuration object should actually be immutable and this not have any setters. Looking at `OSGiUpnpServiceConfiguration`, I'd actually say that it itself should not hold a `configuration` object as a field, since it itself is already the configuration.
The construction of the `StreamClientConfiguration` could actually be done in `createStreamClient()`, since this seems to be the only place, where it is used. And there, it can easily construct a `StreamClientConfiguration` through its constructor.

morph166955 replied:
The problem is that we have a chicken before the egg.  The creation of StreamClientConfiguration requires an executor service to be specified, or nothing at all (which is what we specify so we don't have an NPE issue).  The executor services aren't created until after the configuration read to determine things like number of threads to create.  I suppose I could create a constructor that does not have an executor service and call it at the end of reading the configuration, but that too is problematic because we would have to call back defaults from StreamClientConfiguration which again is chicken and the egg.  I agree it's not the most elegant, but by creating an empty/default StreamClientConfiguration, and then only updating those configs that are actually changed by the end user, we reduce the overall risk that something isn't going to pull in properly and misbehave.

morph166955 replied:
@kaikreuzer Thoughts? I poked at this a little earlier today and while I don't disagree with your point I think that using a constructor adds complexity.  There are several chicken-before-the-egg problems we would have to address depending on where we put it.

kaikreuzer reviewed (changes requested):
Many thanks @morph166955 and sorry for the long silence.
I have a few small last change requests, then we should be good to merge.

morph166955 commented:
Easy enough.  I'll see if I can throw something in shortly.

kaikreuzer commented:
Hi @morph166955, sorry for the long silence.
I had to re-activate my old computer and get a recent dev environment for JUPnP up and running to better support here. I have looked into it and still believe it all works without setter methods - I have created https://github.com/morph166955/jupnp/pull/3 and I have this version currently running on my production environment. It so far looks good to me.

morph166955 commented:
@kaikreuzer Sorry, I moved to a new home and my dev systems are completely down.  I merged your commit but I'm getting a conflict.  The missing line is what we fixed in #135.  I have no way to resolve the conflict, can you from your side?
